tools/lsp/server_loop.rs

```diff
@@ -352,20 +390,56 @@ pub async fn reload_document(
     Ok(())
 }
 
-/// return the token, and the offset within the file
-fn token_descr(
+fn get_document_and_offset(
     document_cache: &mut DocumentCache,
     text_document: lsp_types::TextDocumentIdentifier,
     pos: Position,
-) -> Option<(SyntaxToken, u32)> {
+) -> Option<(&i_slint_compiler::object_tree::Document, u32)> {
     let o = document_cache.newline_offsets.get(&text_document.uri)?.get(pos.line as usize)?
         + pos.character as u32;
 
     let doc = document_cache.documents.get_document(&text_document.uri.to_file_path().ok()?)?;
-    let node = doc.node.as_ref()?;
-    if !node.text_range().contains(o.into()) {
-        return None;
+    doc.node.as_ref()?.text_range().contains(o.into()).then(|| (doc, o))
+}
+
+fn element_contains(element: &i_slint_compiler::object_tree::ElementRc, offset: u32) -> bool {
+    element.borrow().node.as_ref().map(|n| n.text_range().contains(offset.into())).unwrap_or(false)
+}
+
+fn element_at_position(
+    document_cache: &mut DocumentCache,
+    text_document: lsp_types::TextDocumentIdentifier,
+    pos: Position,
+) -> Result<i_slint_compiler::object_tree::ElementRc, String> {
+    let (doc, offset) = get_document_and_offset(document_cache, text_document, pos)
+        .ok_or("Document not found.".to_string())?;
+
+    let root_component = &doc.root_component;
+    let mut element = root_component.root_element.clone();
+    let mut result = None;
+
+    while element_contains(&element, offset) {
+        result = Some(element.clone());
+        if let Some(c) =
+            element.clone().borrow().children.iter().find(|c| element_contains(c, offset))
+        {
+            element = c.clone();
+        } else {
+            return Ok(result.expect("Result was set in this iteration"));
+        }
     }
+    result.ok_or(format!("No element found at offset {}.", offset))
```

Comment:
The more I think about it, the more I think it would be better to find the `ElementRc` by exact match, because you can trivially find the correct element and you can get the uncompiled bindings. So given the SyntaxToken that `token_descr` returns, traverse upwards until you find a `SyntaxNode` with `kind() == SyntaxKind::Element`. That way you've immediately found the correct SyntaxNode.

I think from that CST you can get the uncompiled bindings using `Bindings()` (which you can't get out of the `ElementRc`).

And for the correct property declarations you could take them from the CST, but since you also need the ones from the base type, locating the `ElementRc` that has this resolve might make sense - so take the span and look for the `ElementRc` in the `root_component` by equality.

Alternatively, creating the `ElementRc` on the fly might just be as good.

Reply:
> So given the `SyntaxToken` that `token_descr` returns, traverse upwards until you find a `SyntaxNode` with `kind() == SyntaxKind::Element`. That way you've immediately found the correct `SyntaxNode`.

That is exactly what I tried to do at first. I ended up having to parse things that are readily available in `Element`. So I tried to find the `Element` from the node, which is pretty much code similar to this.

Creating a fresh `Eleemnt` from the `SyntaxNode` using the `From` trait was my next idea: That involved either finding the `Type` of the parent, which in turn involved finding the parent `Element`. So I went searching for an `Element` again:-)

So I tried searching for the right element and work from there.

The `BindingExpression` in the `Element`'s `BindingsMap` has a `span`, with the offset in the file. I use that to find the `SyntaxToken` that is at that offset and grab the `text_range` from that. I could probably also grab the actual text from there? I thought the client gan get that by itself looking at the text range, so I left that out for the time being:-)

Reply:
> Your approach:
>
> 1. Enter the CST at the root, token_at_offset dives into the tree and returns the SyntaxToken where the cursor is.

Actually I walk the _AST_ to find the Element the furthest from the ASTs root that still contains the cursor.

> 2. You use token's offset, enter the AST at the root and find the ElementRc that most closely encapsulates the offset.
> 3. From that ElementRc in the AST you go through the bindings and for each binding get an offset.

Yes.

> 4. You enter the CST at the root and dive into the CST again to return the `SyntaxNode` for the binding at the given offset.

I enter the CST at the `SyntaxNode` of the `Element`. That is the charm of starting from the Element: It refernces back to its `SyntaxNode`.

The Binding should be not too far from that one ;-) Can I actually use `Bindinng()` straight on the `SyntaxNode` of the `Element`?

> My suggestion:
>
> 1. Enter the CST at the root, token_at_offset dives into the tree and returns the SyntaxToken where the cursor is.
> 2. From the `SyntaxToken` go upwards in the CST to find the `SyntaxNode` with `kind() == Element`. This is literally what my eyes would do if I were looking for the element that surrounds the cursor.
> 3. You can fetch the bindings via `Bindings()` and for each binding you have the `syntax_element::Expression` right there.

That works. But I think I need to `Element` to figure out known properties, etc. So I need to find the `Element` at some point.

Reply:
> > Your approach:
> > 
> > 1. Enter the CST at the root, token_at_offset dives into the tree and returns the SyntaxToken where the cursor is.
> 
> Actually I walk the _AST_ to find the Element the furthest from the ASTs root that still contains the cursor.

Ah yes, that's true. That's why you're not calling `token_descr` after all - you never enter the CST at that state.

> I enter the CST at the `SyntaxNode` of the `Element`. That is the charm of starting from the Element: It refernces back to its `SyntaxNode`.
> 
> The Binding should be not too far from that one ;-) Can I actually use `Bindinng()` straight on the `SyntaxNode` of the `Element`?

Yes, that should work, too - I think.